### PR TITLE
CSP-489 Implements onboarding regions view

### PR DESCRIPTION
**Pull request by YuvrajPatil-DFE**
Implemented Regions controller and an empty CurrentJobTitle view to redirect to it on click on back button on Regions view

### CODE REVIEW

File: src/SFA.DAS.ApprenticeAan.Web/Controllers/Onboarding/RegionController.cs [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+    {
+        _sessionService = sessionService;
+        _regionService = regionService;
+        _validator = validator;
+    }
+
+    [HttpGet]
+    public async Task<IActionResult> Get()
+    {
+        var model = new RegionViewModel
```
Comment (DevenKShah) [inline src/SFA.DAS.ApprenticeAan.Web/Controllers/Onboarding/RegionController.cs:31]: You should restore previous selection of regions if any from session model.

File: src/SFA.DAS.ApprenticeAan.Web/Controllers/Onboarding/RegionController.cs [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+
+        var model = new RegionViewModel()
+        {
+            BackLink = Url.RouteUrl(@RouteNames.Onboarding.CurrentJobTitle)!,
+            Regions = await _regionService.GetRegions()
+        };
+
+        ValidationResult result = _validator.Validate(submitmodel);
+
+        if (!result.IsValid)
```
Comment (DevenKShah) [inline src/SFA.DAS.ApprenticeAan.Web/Controllers/Onboarding/RegionController.cs:53]: If the result is not valid then it is required that we add the errors to the model state, so the validation errors can show on the page. I don't see where this is happening?

**Review by DevenKShah — CHANGES_REQUESTED** [review]

File: src/SFA.DAS.ApprenticeAan.Web/Views/Onboarding/CurrentJobTitle.cshtml [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+
+<div class="govuk-grid-row">
+    <div class="govuk-grid-column-two-thirds">
+        <h1 class="govuk-heading-l">What is your current job title?</h1>
+
+        <p>
+            <input type="text" class="govuk-input" />
+        </p>
+
+
```
[the remark this answers is not in the document]
Comment (YuvrajPatil-DFE) [reply src/SFA.DAS.ApprenticeAan.Web/Views/Onboarding/CurrentJobTitle.cshtml:15]: Implemented

File: src/SFA.DAS.ApprenticeAan.Web/Views/Shared/_ValidationSummary.cshtml [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+                        IEnumerable<ModelMetadata> properties = property.Metadata.Properties.Where(metadata => metadata.ShowForDisplay).OrderBy(metadata => metadata.Order);
+                        var propertyName = property.Metadata.PropertyName;
+                        var state = ViewData.ModelState[propertyName!];
+                        var errors = state?.Errors ?? null;
+                        if (errors != null)
+                        {
+                            foreach (var error in errors)
+                            {
+                                                    <li>
+                                                        <a href="#@propertyName">
```
[the remark this answers is not in the document]
Comment (YuvrajPatil-DFE) [reply src/SFA.DAS.ApprenticeAan.Web/Views/Shared/_ValidationSummary.cshtml:24]: Fixed

File: src/SFA.DAS.ApprenticeAan.Web/Controllers/Onboarding/RegionController.cs [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+using SFA.DAS.ApprenticeAan.Web.Filters;
+using SFA.DAS.ApprenticeAan.Web.Infrastructure;
+using SFA.DAS.ApprenticeAan.Web.Models;
+using SFA.DAS.ApprenticeAan.Web.Models.Onboarding;
+
+namespace SFA.DAS.ApprenticeAan.Web.Controllers.Onboarding;
+
+[Route("onboarding/regions", Name = RouteNames.Onboarding.Regions)]
+[RequiredSessionModel(typeof(OnboardingSessionModel))]
+public class RegionController : Controller
```
[the remark this answers is not in the document]
Comment (YuvrajPatil-DFE) [reply src/SFA.DAS.ApprenticeAan.Web/Controllers/Onboarding/RegionController.cs:14]: Changed

File: src/SFA.DAS.ApprenticeAan.Web/Controllers/Onboarding/CurrentJobTitleController.cs [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+namespace SFA.DAS.ApprenticeAan.Web.Controllers.Onboarding
+{
+    [ExcludeFromCodeCoverage]
+    [Route("onboarding/current-job-title", Name = RouteNames.Onboarding.CurrentJobTitle)]
+    [RequiredSessionModel(typeof(OnboardingSessionModel))]
+    public class CurrentJobTitleController : Controller
+    {
+        public const string ViewPath = "~/Views/Onboarding/CurrentJobTitle.cshtml";
+
+        public CurrentJobTitleController()
```
[the remark this answers is not in the document]
Comment (YuvrajPatil-DFE) [reply src/SFA.DAS.ApprenticeAan.Web/Controllers/Onboarding/CurrentJobTitleController.cs:16]: Removed

File: src/SFA.DAS.ApprenticeAan.Application.UnitTests/Services/RegionServiceTests.cs [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+namespace SFA.DAS.ApprenticeAan.Application.UnitTests.Services;
+
+[TestFixture]
+public class RegionServiceTests
+{
+    [MoqAutoData]
+    public async Task Post_SetsSelectedRegionInOnBoardingSessionModel(
+      Mock<IOuterApiClient> outerApiClient)
+    {
+        var regionService = new RegionService(outerApiClient.Object);
```
[the remark this answers is not in the document]
Comment (YuvrajPatil-DFE) [reply src/SFA.DAS.ApprenticeAan.Application.UnitTests/Services/RegionServiceTests.cs:16]: Renamed

File: src/SFA.DAS.ApprenticeAan.Web/Controllers/Onboarding/RegionController.cs (hunk 1 above)
Comment (YuvrajPatil-DFE) [reply src/SFA.DAS.ApprenticeAan.Web/Controllers/Onboarding/RegionController.cs:31]: Implemented

File: src/SFA.DAS.ApprenticeAan.Application.UnitTests/Services/RegionServiceTests.cs [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+
+        var result = await regionService.GetRegions();
+        result.Should().NotBeNullOrEmpty();
```
[the remark this answers is not in the document]
Comment (YuvrajPatil-DFE) [reply src/SFA.DAS.ApprenticeAan.Application.UnitTests/Services/RegionServiceTests.cs:19]: Implemented

[the remark this answers is not in the document]
Comment (YuvrajPatil-DFE) [reply src/SFA.DAS.ApprenticeAan.Application.UnitTests/Services/RegionServiceTests.cs:19]: Implemented

File: src/SFA.DAS.ApprenticeAan.Web/Controllers/Onboarding/RegionsController.cs [hunk 9]
```diff
@@ -1,28 +1,70 @@
...
     [HttpGet]
     public async Task<IActionResult> Get()
     {
-        var regions = await _regionService.GetRegions();
-        return Ok(regions);
+        var model = new RegionsViewModel
+        {
+            BackLink = Url.RouteUrl(RouteNames.Onboarding.CurrentJobTitle)!,
+            Regions = await _regionService.GetRegions()
+        };
```
Comment (DevenKShah) [inline src/SFA.DAS.ApprenticeAan.Web/Controllers/Onboarding/RegionsController.cs:36]: Repetitive code, should be a private method

File: src/SFA.DAS.ApprenticeAan.Web.UnitTests/Controllers/Onboarding/RegionsControllerTests/RegionsControllerGetTests.cs [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+    public async Task Get_ViewResult_HasCorrectViewPath([Greedy] RegionsController sut)
+    {
+        sut.AddUrlHelperMock();
+        var result = await sut.Get();
+
+        result.As<ViewResult>().ViewName.Should().Be(RegionsController.ViewPath);
+    }
+
+    [MoqAutoData]
+    public async Task Get_ViewModel_HasBackLink(
```
Comment (DevenKShah) [inline src/SFA.DAS.ApprenticeAan.Web.UnitTests/Controllers/Onboarding/RegionsControllerTests/RegionsControllerGetTests.cs:34]: Add coverage for getting regions is missing

File: src/SFA.DAS.ApprenticeAan.Application/Services/RegionsService.cs
Comment (DevenKShah) [inline src/SFA.DAS.ApprenticeAan.Application/Services/RegionsService.cs:1]: The interface needs renaming as well to IRegionsService

File: src/SFA.DAS.ApprenticeAan.Web.UnitTests/Controllers/Onboarding/RegionsControllerTests/RegionsControllerPostTests.cs [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+using SFA.DAS.Testing.AutoFixture;
+
+namespace SFA.DAS.ApprenticeAan.Web.UnitTests.Controllers.Onboarding.RegionsControllerTests;
+
+[TestFixture]
+public class RegionsControllerPostTests
+{
+    [MoqAutoData]
+    public async Task Post_SetsSelectedRegionInOnBoardingSessionModel(
+  [Frozen] Mock<ISessionService> sessionServiceMock,
```
Comment (DevenKShah) [inline src/SFA.DAS.ApprenticeAan.Web.UnitTests/Controllers/Onboarding/RegionsControllerTests/RegionsControllerPostTests.cs:21]: alignment, please change your Visual studio setting to fix alignment on save

File: src/SFA.DAS.ApprenticeAan.Web.UnitTests/Controllers/Onboarding/RegionsControllerTests/RegionsControllerPostTests.cs [hunk 12]
```diff
@@ -0,0 +1,51 @@
...
+
+        validatorMock.Setup(v => v.Validate(submitmodel)).Returns(validationResult);
+
+        await sut.Post(submitmodel);
+        sessionServiceMock.Verify(s => s.Set(It.Is<OnboardingSessionModel>(m => m.RegionId == submitmodel.SelectedRegionId)));
+    }
+
+    [MoqAutoData]
+    public async Task Post_Errors_WhenSelectedRegionIsNull(
+  [Frozen] Mock<ISessionService> sessionServiceMock,
```
Comment (DevenKShah) [inline src/SFA.DAS.ApprenticeAan.Web.UnitTests/Controllers/Onboarding/RegionsControllerTests/RegionsControllerPostTests.cs:37]: alignment

File: src/SFA.DAS.ApprenticeAan.Web/Validators/Onboarding/RegionSubmitModelValidator.cs
Comment (DevenKShah) [inline src/SFA.DAS.ApprenticeAan.Web/Validators/Onboarding/RegionSubmitModelValidator.cs:1]: Rename file to RegionsSubmitModelValidator

**Review by DevenKShah — COMMENTED** [review]
Few minor fixes required

File: src/SFA.DAS.ApprenticeAan.Web.UnitTests/Controllers/Onboarding/RegionsControllerTests/RegionsControllerPostTests.cs (hunk 11 above)
Comment (YuvrajPatil-DFE) [reply src/SFA.DAS.ApprenticeAan.Web.UnitTests/Controllers/Onboarding/RegionsControllerTests/RegionsControllerPostTests.cs:21]: Redone the alignment manually.  I do have the same settings, somehow its now applying the way we want

File: src/SFA.DAS.ApprenticeAan.Web.UnitTests/Controllers/Onboarding/RegionsControllerTests/RegionsControllerGetTests.cs (hunk 10 above)
Comment (YuvrajPatil-DFE) [reply src/SFA.DAS.ApprenticeAan.Web.UnitTests/Controllers/Onboarding/RegionsControllerTests/RegionsControllerGetTests.cs:34]: Implemented. Verified Backlink, Regions list and SelectedRegion if any.

**Review by DevenKShah — APPROVED** [review]

**Review by DevenKShah — APPROVED** [review]

**Pull request merged**